File: trunk/java/log_client_java/src/com/persiste/sdk/HttpClientWrapper.java
/**
* Persiste. Java SDK
*
* The cloud based logging platform.
*
* @package             Persiste. SDK
* @author              Evance Soumaoro
* @copyright           Copyright (c) 2012 - 20xx, Evansofts.
* @license             Evansofts Proprietary Licence (EPL)
* @link                http://evansofts.com
* @version             Version 0.1
*/

package com.persiste.sdk;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import org.apache.http.HttpResponse;
import org.apache.http.NameValuePair;
import org.apache.http.client.HttpClient;
import org.apache.http.client.entity.UrlEncodedFormEntity;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.impl.client.DefaultHttpClient;

import com.persiste.sdk.json.JSONObject;
import com.persiste.sdk.json.JSONValue;


public class HttpClientWrapper {
	private static INotifiable notification_receiver;
	private static boolean is_async;
	
	private String response;
	private String url;
	private boolean has_error;
	
	public HttpClientWrapper(String url) {
		this.response="";
		this.url=url;
		this.has_error=false;
	}
	
	public void get() {
		HttpGetAction action=new HttpGetAction(this.url);
		if(HttpClientWrapper.is_async==true){
			Thread thread = new Thread(action);
			thread.start();
		}else{
			action.run();
		}
	} 
		
	public void post( ArrayList<NameValuePair> data) {
		HttpPostAction action=new HttpPostAction(this.url, data);
		if(HttpClientWrapper.is_async==true){
			Thread thread = new Thread(action);
			thread.start();
		}else{
			action.run();
		}
	} 
	
	public boolean hasError(){
		return this.has_error;		
	}
		
	protected JSONObject packData(HashMap<String, Object> data){
		return null;
	}
	
	@SuppressWarnings("unchecked")
	protected LogserviceResponse unpackData(String response){
		LogserviceResponse formated_response=null;
		if(response.equals("404")){
			formated_response=new LogserviceResponse("failure","Missing required parameter");
			return formated_response;
		}
		
		if(response.equals("401")){
			formated_response=new LogserviceResponse("failure","Authentication error, check your credentials");
			return formated_response;
		}
		
		try{
			JSONObject json=(JSONObject)JSONValue.parse(response);
			formated_response=new LogserviceResponse((String)json.get("status"),(String)json.get("message"));
			ArrayList<JSONObject> logs=(ArrayList<JSONObject>)json.get("data");
			if(logs!=null){
				ArrayList<Object> data=new ArrayList<Object>();
				Log log=null;
				for(JSONObject obj:logs){
					log=new Log(Integer.parseInt((String)obj.get("level")),(String)obj.get("title"),(String) obj.get("description"), (String )obj.get("emiter_email"), (Map<String, String>) obj.get("custom_fields") , (String) obj.get("stack_trace"));
					data.add(log);
				}
				formated_response.setData(data);
			}	
			
		}catch (Exception e) {
			e.printStackTrace();
		}
		
		return formated_response;
	}
	
	public static void setParams(INotifiable  notification_receiver,boolean is_async){
		HttpClientWrapper.notification_receiver=notification_receiver;
		HttpClientWrapper.is_async=is_async;
	}
		
	private class HttpGetAction implements Runnable{
		private String url;
	
		public HttpGetAction(String url) {
			this.url=url;
		}
		
		@Override
		public void run() {
			try {
			    DefaultHttpClient client = new DefaultHttpClient();
				HttpGet httpclient = new HttpGet(this.url);
		        HttpResponse http_response = client.execute(httpclient);
		        InputStream content = http_response.getEntity().getContent();
				BufferedReader buffer = new BufferedReader(new InputStreamReader(content));
				String s = "";
				while ((s = buffer.readLine()) != null) {
					HttpClientWrapper.this.response += s;
				}
		    } catch (Exception e) { 
		    	HttpClientWrapper.this.response="{\"status\":\"failure\",\"message\":\"Http client error check connection\"}";
		    	HttpClientWrapper.this.has_error=true;
		    }

			if(HttpClientWrapper.notification_receiver!=null){
				LogserviceResponse formated_response=HttpClientWrapper.this.unpackData(HttpClientWrapper.this.response);
				HttpClientWrapper.notification_receiver.LoggingClientCompleted(formated_response,HttpClientWrapper.this.has_error);
			}
			
			
		}
		
	}
		
	private class HttpPostAction implements Runnable{
		private String url;
		private ArrayList<NameValuePair> data;

		public HttpPostAction(String url,ArrayList<NameValuePair> data) {
			this.url=url;
			this.data=data;
		}
		
		@Override
		public void run() {
			try {
			    	HttpClient httpclient = new DefaultHttpClient();
				    HttpPost httppost = new HttpPost(this.url);
			        httppost.setEntity(new UrlEncodedFormEntity(this.data));
			        HttpResponse http_response = httpclient.execute(httppost);
			        InputStream content = http_response.getEntity().getContent();
					BufferedReader buffer = new BufferedReader(new InputStreamReader(content));
					String s = "";
				while ((s = buffer.readLine()) != null) {
					HttpClientWrapper.this.response += s;
				}
			} catch (Exception e) {  
				HttpClientWrapper.this.response="{\"status\":\"failure\",\"message\":\"Http client error check connection\"}";
				HttpClientWrapper.this.has_error=true;
			}

			if(HttpClientWrapper.notification_receiver!=null){
				LogserviceResponse formated_response=HttpClientWrapper.this.unpackData(HttpClientWrapper.this.response);
				HttpClientWrapper.notification_receiver.LoggingClientCompleted(formated_response,HttpClientWrapper.this.has_error);
			}
		}

	}
	
}

File: csharp/log_client_cs/PersisteDotSdk/HttpClientWrapper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Newtonsoft.Json;
using System.Net;
using System.Threading;

namespace PersisteDotSdk {
    
    public class HttpClientWrapper {
        private static INotifiable notification_receiver;
	    private static bool is_async;
	

        public String url; 
	
	    public HttpClientWrapper(String url) {
		    this.url=url;
	    }
	
	    public void get() {
            HttpGetAction action=new HttpGetAction(this,this.url);
		    if(HttpClientWrapper.is_async==true){
			    Thread thread = new Thread(action.Run);
			    thread.Start();
		    }else{
			    action.Run();
		    }
	    } 
		
	    public void post( Dictionary<String, String> data) {
		    HttpPostAction action=new HttpPostAction(this,this.url, data);
		    if(HttpClientWrapper.is_async==true){
			    Thread thread = new Thread(action.Run);
			    thread.Start();
		    }else{
			    action.Run();
		    }
	    } 
	
	    public String packData(Dictionary<String, Object> data){
		    return null;
	    }

        public LogServiceResponse unpackData(String response) {
            LogServiceResponse formated_response = null;
		    if( String.Equals(response,"404") ){
                formated_response= new LogServiceResponse("failure","Missing required parameter");
			    return formated_response;
		    }

            if (String.Equals(response, "401")) {
                formated_response = new LogServiceResponse("failure", "Authentication error, check your credentials");
                return formated_response;
		    }
            
            try { 
                formated_response=JsonConvert.DeserializeObject<LogServiceResponse>(response);

                if (formated_response.Data != null) {
                    Log log = null;
                    int level;
                    String title;
                    String description;
                    String emiter;
                    Dictionary<String, String> custom_fields;
                    String stack_trace;

                    List<Object> log_list = new List<object>();
                    Newtonsoft.Json.Linq.JToken value;
                    foreach (System.Collections.Generic.ICollection<System.Collections.Generic.KeyValuePair<System.String, Newtonsoft.Json.Linq.JToken>> item in formated_response.Data) {
                        var p = item.ToDictionary((keyItem) => keyItem.Key, (valueItem) => valueItem.Value);
                        value = (p.TryGetValue("level", out value)) ? value : null;
                        level = Int16.Parse((String)value);
                        value = (p.TryGetValue("title", out value)) ? value : null;
                        title = (String)value;
                        value = (p.TryGetValue("description", out value)) ? value : null;
                        description = (String)value;
                        value = (p.TryGetValue("emiter_email", out value)) ? value : null;
                        emiter = (String)value;
                        //we dont send back custome field for now
                        //value = (p.TryGetValue("custom_fields", out value)) ? value : null;
                        custom_fields = null;
                        value = (p.TryGetValue("stack_trace", out value)) ? value : null;
                        stack_trace = (String)value;

                        log = new Log(level, title, description, emiter, custom_fields, stack_trace);
                        log_list.Add(log);

                    }
                    formated_response.Data = log_list;
                }

            }catch(Exception){}

            return formated_response;
	    }

   
        public static bool isAsync(){
		    return HttpClientWrapper.is_async;		
	    }

        public static INotifiable getNotificationReceiver() {
            return HttpClientWrapper.notification_receiver;
        }

	    public static void setParams(INotifiable  notification_receiver,bool is_async){
		    HttpClientWrapper.notification_receiver=notification_receiver;
		    HttpClientWrapper.is_async=is_async;
	    }
		 
    }


    



}
